README
# FDC-SDS Serial Disk Server for the Altair FDC+

FDC-SDS is a Linux implentation of the Serial Drive Communications Protocol (https://deramp.com/downloads/altair/hardware/fdc+/FDC%20Serial%20Server%20Protocol.txt) for the FDC+ Enhanced Floppy Disk Controller for the Altair 8800 (https://deramp.com/fdc_plus.html).

## Installation
./configure
make
make install

## Baud Rate
The FDC+ serial drive can operate at one of three baud rates: 403.2K, 460.8K, or 230.4K baud. FDC-SDS only supports 230.4K (default) and 460.8K.

If the baud rate is changed in the FDC-SDS server, the corresponding change must also be made on the FDC+. Baud rates are changed in the FDC+ using the monitor.

## Disk Images
Numerous disk image files of original Altair software are available via links on the FDC+ web page. Images files have a ".dsk" extension. You can also recognize disk images by their file size. Eight inch disk images are 330K, Minidisk images are 75K.

The FDC-SDS server is indifferent to whether an eight inch image or Minidisk image is loaded, however, the image type loaded must match drive type selected on the FDC+.

## Example
fdcsds -p /dev/ttyUSB0 -0 cpm22.dsk


ChangeLog
2020-02-29  Patrick Linstruth <patrick@deltecent.com>
        * Initial 1.0 Release


io.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <termios.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/select.h>
#include "io.h"
#include "display.h"

static int fd;

static struct termios oldtio;
static struct termios newtio;

static char debug[80];

int openPort(char *device, int baud)
{
	if (device == NULL) {
		return (-1);
	}

	if ((fd = open(device, O_RDWR | O_NOCTTY | O_NONBLOCK)) == -1) {
		return fd;
	}

	tcgetattr(fd, &oldtio); /* save current serial port settings */

	memset(&newtio, 0, sizeof(newtio)); /* clear struct for new port settings */
	newtio.c_cflag = CLOCAL | CS8 | CREAD;
	newtio.c_iflag = 0;
	newtio.c_oflag = 0;

	/*
	** The FDC+ only supports 230.4K, 403.2K and 460.8K
	** Linux does not appear to support 403.2K
	*/
	switch (baud) {
		case 460800:
			cfsetspeed(&newtio, B460800);
			break;

		default:
			baud = 230400;
			cfsetspeed(&newtio, B230400);
			break;
	}

	tcsetattr(fd, TCSANOW, &newtio);

	tcflush(fd, TCIFLUSH);

	displayPort(device);
	displayBaud(baud);

	return fd;
}

int closePort()
{
	tcsetattr(fd, TCSANOW, &oldtio);

	return close(fd);
}

int recvByte(uint8_t *byte, int tsecs)
{
	int n;
	fd_set input;
	struct timeval timeout;

	FD_ZERO(&input);
	FD_SET(fd, &input);

	timeout.tv_sec = tsecs;
	timeout.tv_usec = 0;

	n = select(fd+1, &input, NULL, NULL, &timeout);

	if (n < 0) {		/* ERROR */
		displayError("SELECT", errno);

		return (-1);
	}
	else if (n==0) {	/* TIMEOUT */
		return (0);
	}

	n = read(fd, byte, 1);

	return(n);
}

int recvBuf(void *buffer, int length, int tsecs)
{
	int count = 0;
	int bytes;
	uint8_t lsb;
	uint8_t msb;
	uint16_t chksum;
	/*
	** Receive requested length
	*/
	while (length) {

		bytes = recvByte(buffer + count, tsecs);

		if (bytes > 0) {
			count += bytes;
			length -= bytes;
		}
		else {
			return(count);
		}
	}

	/*
	** Receive Checksum
	*/
	recvByte(&lsb, 1);
	recvByte(&msb, 1);
	chksum = (msb << 8) + lsb;

	if (chksum != calcChksum(buffer, count)) {
		displayError("CHECKSUM", 0);

		count = 0;
	}

	return count;
}

/*
** Send buffer and append checksum
*/
int sendBuf(void *buffer, int length, int tsecs)
{
	uint16_t chksum;
	uint8_t msb;
	uint8_t lsb;
	int bytes;
	int sent = 0;
	int n;
	fd_set output;
	struct timeval timeout;

	FD_ZERO(&output);
	FD_SET(fd, &output);

	timeout.tv_sec = tsecs;
	timeout.tv_usec = 0;

	while (sent < length) {
		n = select(fd+1, NULL, &output, NULL, &timeout);

		if (n == -1) {
			displayError("SELECT", errno);
			return -1;
		}

		bytes = write(fd, buffer + sent, length - sent);

		if (bytes == 0) {
			displayError("WRITE TIMEOUT", errno);
			return -1;
		}
		else if (bytes == -1) {
			displayError("WRITE", errno);
			return -1;
		}

		sent += bytes;
	}

	chksum = calcChksum(buffer, length);

	lsb = chksum & 0x00ff;

	bytes = write(fd, &lsb, 1);

	if (bytes != 1) {
		displayError("WRITE", errno);
		return -1;
	}

	msb = (chksum & 0xff00) >> 8;

	bytes = write(fd, &msb, 1);

	if (bytes != 1) {
		displayError("WRITE", errno);
		return -1;
	}

	tcdrain(fd);

	return 0;
}

uint16_t calcChksum(void *buffer, int length)
{
	uint16_t chksum = 0;

	while (length--) {
		chksum += * (uint8_t *) buffer++;
	}

	return chksum;
}




drive.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/select.h>
#include "fdcsds.h"
#include "drive.h"
#include "display.h"

/*
** 137 bytes per sector, 32 sectors per track, 77 tracks
*/

drvstat_t drvstat[MAX_DRIVES];
uint8_t trackbuf[MAX_TRACK_LEN];
int fdc_errno;

int mountDrive(int drive, char *filename)
{
	if (drive >= MAX_DRIVES) {
		return -1;
	}

	strncpy(drvstat[drive].filename, filename, MAX_PATH);

	if ((drvstat[drive].fd = open(filename, O_RDWR)) == -1) {
		drvstat[drive].mounted = FALSE;
		displayMount(drive, "--ERROR--");
		displayError("MOUNT", errno);
	}
	else {
		drvstat[drive].mounted = TRUE;
		displayMount(drive, filename);
	}

	return drvstat[drive].fd;
}

int unmountDrive(int drive)
{
	if (drive >= MAX_DRIVES) {
		return -1;
	}

	if (drvstat[drive].mounted && drvstat[drive].fd != -1) {
		close(drvstat[drive].fd);

		drvstat[drive].mounted = FALSE;
		drvstat[drive].fd = -1;

		displayMount(drive, NULL);
	}

	return 0;
}

void unmountAll()
{
	int d;

	for (d = 0; d < MAX_DRIVES; d++) {
		unmountDrive(d);
	}
}

int writeProtect(int drive, int flag)
{
	if (drive >= MAX_DRIVES) {
		return -1;
	}

	if (drvstat[drive].mounted) {
		drvstat[drive].readonly = (flag) ? TRUE : FALSE;

		displayRO(drive, flag);
	}

	return (0);
}

int readTrack(int drive, int track, int length, void *buffer)
{
	off_t offset;
	int bytes;

	if (drive >= MAX_DRIVES) {
		displayError("INVALID DRIVE", 0);
		return -1;
	}

	if (drvstat[drive].mounted && drvstat[drive].fd == -1) {
		displayError("DISK NOT MOUNTED", 0);
		return -1;
	}

	offset = track * length;

	displayTrack(drive, track);

	/*
	** Seek to track
	*/
	if (lseek(drvstat[drive].fd, offset, SEEK_SET) != offset) {
		displayError("LSEEK", errno);
		return(-1);
	}

	/*
	** Read track
	*/
	if ((bytes = read(drvstat[drive].fd, buffer, length)) != length) {
		displayError("READ TRACK", errno);
	}

	return (bytes);
}

int writeTrack(int drive, int track, int length, void *buffer)
{
	off_t offset;
	int bytes;

	if (drive >= MAX_DRIVES) {
		displayError("INVALID DRIVE", 0);
		fdc_errno = FDC_NOT_READY;
		return (-1);
	}

	if (!drvstat[drive].mounted || drvstat[drive].fd == -1) {
		displayError("DISK NOT MOUNTED", 0);
		fdc_errno = FDC_NOT_READY;
		return (-1);
	}

	if (drvstat[drive].readonly) {
		displayError("DISK READ ONLY", 0);
		fdc_errno = FDC_WRITE_ERR;
		return (-1);
	}

	offset = track * length;

	displayTrack(drive, track);

	/*
	** Seek to track
	*/
	if (lseek(drvstat[drive].fd, offset, SEEK_SET) != offset) {
		displayError("LSEEK", errno);
		fdc_errno = FDC_WRITE_ERR;
		return (-1);
	}

	/*
	** Write track
	*/
	if ((bytes = write(drvstat[drive].fd, buffer, length)) != length) {
		displayError("WRITE TRACK", errno);
		fdc_errno = FDC_WRITE_ERR;
		return (-1);
	}

	fdc_errno = FDC_OK;

	return (bytes);
}



drive.h
#ifndef	_DRIVE_H
#define	_DRIVE_H

#include <stdint.h>

#define	MAX_DRIVES	16
#define	MAX_TRACKS	77
#define	MAX_TRACK_LEN	(137*32)
#define MAX_DISK_SIZE	MAX_TRACK_LEN * MAX_TRACKS
#define	MAX_PATH	128

typedef struct DRVSTAT {
	int		fd;
	char		filename[MAX_PATH];
	uint8_t		mounted;
	uint8_t		readonly;
	uint8_t		hdld;
	uint16_t	track;
} drvstat_t;

#define	LSB(b)		(b & 0xff)
#define	MSB(b)		((b & 0xff00) >> 8)
#define	WORD(lsb,msb)	((msb << 8) | lsb)

#define	FDC_OK		0x00
#define	FDC_NOT_READY	0x01
#define	FDC_CHKSUM_ERR	0x02
#define	FDC_WRITE_ERR	0x03

extern int	fdc_errno;

/*
** FDC+ Command / Response Block
*/
typedef struct CRBLOCK {
	char		cmd[4];

	union {
		struct {
			uint8_t	lsb1;
			uint8_t	msb1;
		};
		uint16_t	param1;
	};

	union {
		struct {
			uint8_t	lsb2;
			uint8_t	msb2;
		};
		uint16_t	param2;
	};
} crblk_t;

extern drvstat_t drvstat[MAX_DRIVES];
extern uint8_t trackbuf[MAX_TRACK_LEN];

extern int mountDrive(int drive, char *filename);
extern int unmountDrive(int drive);
extern void unmountAll();
extern int writeProtect(int drive, int flag);
extern int readTrack(int drive, int track, int length, void *buffer);
extern int writeTrack(int drive, int track, int length, void *buffer);

#endif


display.c
#include <string.h>
#include <ctype.h>
#include <libgen.h>
#include <ncurses.h>
#include <sys/errno.h>
#include "fdcsds.h"
#include "display.h"
#include "drive.h"

#define	MAX_LINE	LINES-1
#define MAX_COL		COLS-1

#define TITLE_LINE	0
#define	TITLE_COL	0

#define COPY_LINE	0
#define	COPY_COL	MAX_COL-sizeof(FDCSDS_COPYRIGHT)+2

#define	PORT_LINE	2
#define PORT_COL	0
#define PORT_TEXT	"PORT:"

#define	BAUD_LINE	PORT_LINE
#define BAUD_COL	PORT_COL+sizeof(PORT_TEXT)+21
#define BAUD_TEXT	"BAUD RATE:"

#define COMMAND_LINE	PORT_LINE
#define COMMAND_COL	BAUD_COL+sizeof(BAUD_TEXT)+10
#define	COMMAND_TEXT	"COMMAND:"

#define	BLOCK_LINE	COMMAND_LINE
#define BLOCK_COL	COMMAND_COL + sizeof(COMMAND_TEXT) + 7

#define	DRIVE_LINE	4
#define	DRIVE_COL	0
#define	DRIVE_TEXT	"Disk -                                Disk Enable -  Head Load -  Track --  RO -"
#define	DRIVE_NUM	5
#define	DRIVE_FILE	8
#define	DRIVE_ENA	50
#define	DRIVE_HEAD	63
#define	DRIVE_TRACK	72
#define	DRIVE_RO	79

#define ERROR_LINE	DRIVE_LINE + 5
#define ERROR_COL	0
#define	ERROR_TEXT	"ERROR:"

#define DEBUG_LINE	ERROR_LINE+1
#define	DEBUG_COL	0

#define	BUFFER_LINE	MAX_LINE-12
#define	BUFFER_COL	0

#define	HELP_LINE	MAX_LINE
#define	HELP_COL	0
#define	HELP_TEXT	"[C] = Clear Error Message | [Q] = Quit Program | [V] = Verbose Toggle"

char dispbuf[128];

void displayInit()
{
	int d;

	initscr();

	LINES = 24;
	COLS = 80;

	raw();
	noecho();
	curs_set(0);
	timeout(0);

	clear();

	move(TITLE_LINE, TITLE_COL);
	printw(FDCSDS_NAME);
	printw(" v");
	printw(VERSION);
	move(COPY_LINE, COPY_COL);
	printw(FDCSDS_COPYRIGHT);

	move(PORT_LINE, PORT_COL);
	printw(PORT_TEXT);

	move(BAUD_LINE, BAUD_COL);
	printw(BAUD_TEXT);

	move(COMMAND_LINE, COMMAND_COL);
	printw(COMMAND_TEXT);

	move(ERROR_LINE, ERROR_COL);
	printw(ERROR_TEXT);

	for (d = 0; d < 4; d++) {
		move(DRIVE_LINE + d, DRIVE_COL);
		printw(DRIVE_TEXT);
		move(DRIVE_LINE + d, DRIVE_NUM);
		printw("%d", d);
	}

	move(HELP_LINE, HELP_COL);
	printw(HELP_TEXT);

	refresh();
}

void displayReset()
{
	endwin();
}

int displayGetch()
{
	return getch();
}

void displayPort(char *port)
{
	move(PORT_LINE, PORT_COL+sizeof(PORT_TEXT));
	printw("%-20.20s", basename(port));
	refresh();
}

void displayBaud(int baud)
{
	move(BAUD_LINE, BAUD_COL+sizeof(BAUD_TEXT));
	printw("%d", baud);
	refresh();
}

void displayCommand(char *command)
{
	move(COMMAND_LINE, COMMAND_COL+sizeof(COMMAND_TEXT));
	printw("%-4.4s", command);
	refresh();
}

void displayBlock(int drive, int track, int length)
{
	move(BLOCK_LINE, BLOCK_COL);
	if (drive >= 0 && drive <= 0xff) {
		printw("D:%02X ", drive);
	}
	else {
		printw("D:-- ");
	}
	if (track != -1) {
		printw("T:%02d ", track);
	}
	else {
		printw("T:-- ");
	}
	if (length != -1) {
		printw("L:%04d", length);
	}
	else {
		printw("L:----");
	}

	refresh();
}

void displayError(char *string, int err)
{
	char error[80];

	strcpy(error, string);

	if (err) {
		strcat(error, " (");
		strcat(error, strerror(err));
		strcat(error, ")");

		/*
		** Reset errno
		*/
		errno = 0;
	}

	move(ERROR_LINE, ERROR_COL+sizeof(ERROR_TEXT));
	printw("%-60.60s", error);
	refresh();
}

void displayDebug(char *string)
{
	move(DEBUG_LINE, DEBUG_COL);
	clrtoeol();
	printw("%s", string);
	refresh();
}

// need to make a displayEnabled

void displayHead(int drive, int head)
{
	int d;

	if (drive < 0 || drive > 3) {
		return;
	}

	for (d = 0; d < 4; d++) {
		move(DRIVE_LINE + d, DRIVE_ENA);
		printw("%c", (d == drive) ? '*' : '-');

		move(DRIVE_LINE + d, DRIVE_HEAD);
		if (d == drive) {
			printw("%c", (head) ? '*' : '-');
		}
		else {
			printw("-");
		}
	}

	refresh();
}

void displayTrack(int drive, int track)
{
	if (drive < 0 || drive > 3) {
		return;
	}

	move(DRIVE_LINE + drive, DRIVE_TRACK);
	printw("%02d", track);

	refresh();
}

void displayMount(int drive, char *path)
{
	if (drive < 0 || drive > 3) {
		return;
	}

	move(DRIVE_LINE + drive, DRIVE_FILE);
	printw("%-25.25s", (path != NULL) ? path : "");

	refresh();
}

void displayRO(int drive, int wp)
{
	if (drive < 0 || drive > 3) {
		return;
	}

	move(DRIVE_LINE + drive, DRIVE_RO);
	printw("%c", (wp) ? '*' : '-');

	refresh();
}

void displayBuffer(char *string, void *buffer, int length)
{
	int i;
	int maxlen;
	int l = BUFFER_LINE;
	int c = BUFFER_COL;

	if (!length) {
		return;
	}

	for (i = BUFFER_LINE; i < BUFFER_LINE + 10; i++) {
		move(i, 5);
		clrtoeol();
	}

	maxlen = (length > 160) ? 160 : length;

	for (i=0; i<maxlen; i++) {
		if (!(i % 16)) {
			if (i) {
				l++;
				c = 0;
			}
			move(l, c);
			printw("%04X: ", i);
		}

		move(l, 6 + (i % 16) * 3);
		printw("%02X ", * (uint8_t *) buffer);
		move(l, 54 + i % 16);
		if (isprint(* (uint8_t *) buffer)) {
			printw("%c", * (uint8_t *) buffer);
		}
		else {
			printw(".");
		}

		buffer++;
	}

	move(BUFFER_LINE+11, BUFFER_COL);
	clrtoeol();
	printw(string);

	refresh();
}


fdcsds.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <errno.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/select.h>
#include "fdcsds.h"
#include "io.h"
#include "drive.h"
#include "display.h"

static void help();
static int processCmd(crblk_t *cmd);
static int statCmd(crblk_t *cmd);
static int readCmd(crblk_t *cmd);
static int writCmd(crblk_t *cmd);
static int writResp(crblk_t *cmd, int resp);

char *port = NULL;
int baud = 230400;
char dbuf[128];

/*
** Flag Variables
*/
int f_help = FALSE;
int f_verbose = FALSE;
int f_debug = FALSE;

struct option longopts[] = {
   { "port",	required_argument,	NULL,		'p' },
   { "baud",	required_argument,	NULL,		'b' },
   { "load0",	required_argument,	NULL,		'0' },
   { "load1",	required_argument,	NULL,		'1' },
   { "load2",	required_argument,	NULL,		'2' },
   { "load3",	required_argument,	NULL,		'3' },
   { "ro",	required_argument,	NULL,		'r' },
   { "help",	no_argument,     	&f_help,	TRUE },
   { "debug",	no_argument,		&f_debug,	TRUE },
   { "verbose",	no_argument,		&f_verbose,	TRUE },
   { 0, 0, 0, 0 }
};

int main(int argc, char **argv)
{
	int running = TRUE;
	int bytes;
	crblk_t	cmd;
	int c;

	memset(&drvstat, 0, sizeof(drvstat));

	displayInit();

	while ((c = getopt_long(argc, argv, ":0:1:2:3:p:b:r:vh", longopts, NULL)) != -1) {
		switch (c) {
			case '0':
			case '1':
			case '2':
			case '3':
				mountDrive(c-'0', optarg);
				break;

			case 'r':
				writeProtect(atoi(optarg), TRUE);
				break;

			case 'p':
				port = optarg;
				break;

			case 'b':
				baud = atoi(optarg);
				break;

			case 'd':
				f_debug = TRUE;
				break;

			case 'v':
				f_verbose = TRUE;
				break;

			case 'h':
				f_help = TRUE;
				break;

			case 0:
				/* getopt_long() set a variable, just keep going */
				break;
		}
	}

	if (f_help) {
		displayReset();
		help();
		exit(1);
	}

	if (port == NULL) {
		displayReset();
		help();
		printf("You must specify a serial port with '-p' option.\n\n");
		exit(1);
	}

	if(openPort(port, baud) == -1) {
		displayReset();
		perror("Unable to open serial port");
		exit(errno);
	}

	/*
	** Command loop
	*/
	while (running) {

		switch(toupper(displayGetch())) {
			case 'C':
				displayError("", 0);
				break;

			case 'Q':
				running = FALSE;
				break;

			case 'V':
				f_verbose = !f_verbose;
				break;

			default:
				break;
		}

		/*
		** Wait for command from FDC+
		*/
		bytes = recvBuf(&cmd, sizeof(cmd), 1);

		if (bytes && f_verbose) {
			displayBuffer("", &cmd, sizeof(cmd));
		}

		if (bytes == sizeof(cmd)) {
			processCmd(&cmd);
		}
		else {
			displayCommand("----");
			displayBlock(-1, -1, -1);
		}
	}

	closePort();
	unmountAll();
	displayReset();
}

static void help()
{
	printf("\n%s v%s\n", FDCSDS_NAME, VERSION);
	printf("%s\n\n", FDCSDS_COPYRIGHT);
	printf("Serial Disk Server compatible with the FDC+ Enhanced Floppy Disk\n");
	printf("Controller for the Altair 8800 available at http://www.deramp.com\n\n");
	printf("server [options] -p port\n\n");
	printf("Options:\n");
	printf("-[0-3] file\tMount disk image file to drive 0-3\n");
	printf("\t\tThe FDC+ supports 330K 8 inch and 75K Minidisk images\n");
	printf("-b baud\t\tSet serial port speed (default=%d)\n", baud);
	printf("-p port\t\tSerial port (required)\n");
	printf("-r [0-3]\tMake drive 0-3 read only\n");
	printf("-v\t\tVerbose display\n\n");
}

static int processCmd(crblk_t *cmd)
{
	if (!strncmp(cmd->cmd, "STAT", 4)) {
		statCmd(cmd);
	}
	else if (!strncmp(cmd->cmd, "READ", 4)) {
		readCmd(cmd);
	}
	else if (!strncmp(cmd->cmd, "WRIT", 4)) {
		writCmd(cmd);
	}

	return -1;
}

static int statCmd(crblk_t *cmd)
{
	int i;
	uint16_t data = 0;
	int drive;

	displayCommand("STAT");

	drive = cmd->lsb1;

	/*
	** Save head load status and track for drive
	*/
	if (drive < MAX_DRIVES) {
		drvstat[drive].hdld = cmd->msb1;
		drvstat[drive].track = WORD(cmd->lsb2, cmd->msb2);

		displayHead(drive, drvstat[drive].hdld);
		displayTrack(drive, drvstat[drive].track);

	}
	else {
		for (i=0; i < MAX_DRIVES; i++) {
			drvstat[i].hdld = FALSE;

			displayHead(drive, drvstat[i].hdld);
		}

		displayBlock(drive, -1, -1);
	}

	for (i=0; i < MAX_DRIVES; i++) {
		data |= (drvstat[i].mounted << i);
	}

	cmd->lsb2 = LSB(data);
	cmd->msb2 = MSB(data);

	if (f_verbose) {
		displayBuffer("", cmd, sizeof(crblk_t));
	}

	return (sendBuf(cmd, sizeof(crblk_t), 5));
}

static int readCmd(crblk_t *cmd)
{
	int length;
	int drive;
	int track;
	int bytes;

	displayCommand("READ");

	drive = cmd->msb1 >> 4;
	track = WORD(cmd->lsb1, cmd->msb1) & 0x0fff;
	length = WORD(cmd->lsb2, cmd->msb2);

	if (f_debug) {
		sprintf(dbuf, "READ TRACK D:%02d T:%02d L:%04d", drive, track, length);
		displayDebug(dbuf);
	}

	if (f_verbose) {
		displayBlock(drive, track, length);
	}

	if (drive < MAX_DRIVES) {
		drvstat[drive].track = track;
	}

	if (readTrack(drive, track, length, trackbuf) != length) {
		return (-1);
	}

	if (f_verbose) {
		displayBuffer("", trackbuf, length);
	}

	return (sendBuf(trackbuf, length, 5));
}

static int writCmd(crblk_t *cmd)
{
	int length;
	int drive;
	int track;

	displayCommand("WRIT");

	drive = cmd->msb1 >> 4;
	track = WORD(cmd->lsb1, cmd->msb1) & 0x0fff;
	length = WORD(cmd->lsb2, cmd->msb2);

	if (f_verbose) {
		displayBlock(drive, track, length);
	}

	if (drive >= MAX_DRIVES) {
		return(writResp(cmd, FDC_NOT_READY));
	}

	drvstat[drive].track = track;

	writResp(cmd, FDC_OK);

	/*
	** Wait for track
	*/
	if (recvBuf(trackbuf, length, 5) == 0) {
		displayError("RECEIVE TIMEOUT", 0);
	}

	if (f_verbose) {
		displayBuffer("", trackbuf, length);
	}

	memcpy(cmd->cmd, "WSTA", 4);

	if (writeTrack(drive, track, length, trackbuf) == length) {
		writResp(cmd, FDC_OK);
	}
	else {
		writResp(cmd, fdc_errno);
		return(-1);
	}

	return(0);
}

static int writResp(crblk_t *cmd, int resp)
{
	cmd->lsb1 = LSB(resp);
	cmd->msb1 = MSB(resp);

	if (f_verbose) {
		displayBuffer("", cmd, sizeof(crblk_t));
	}

	/*
	** Send WRIT response
	*/
	return(sendBuf(cmd, sizeof(crblk_t), 5));
}

